qiskit/pulse/transforms/frames.py

```diff
@@ -0,0 +1,135 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Basic rescheduling functions which take schedule or instructions and return new schedules."""
+
+
+from typing import Dict
+
+from qiskit.pulse.schedule import Schedule
+from qiskit.pulse.resolved_frame import ResolvedFrame, ChannelTracker
+from qiskit.pulse.library import Signal
+from qiskit.pulse.exceptions import PulseError
+from qiskit.pulse import channels as chans, instructions
+from qiskit.pulse.frame import Frame
+
+
+
+def resolve_frames(schedule: Schedule, frames_config: Dict[int, Dict]) -> Schedule:
+    """
+    Parse the schedule and replace instructions on Frames by instructions on the
+    appropriate channels.
+
+    Args:
+        schedule: The schedule for which to replace frames with the appropriate
+            channels.
+        frames_config: A dictionary with the frame index as key and the values are
+            a dict which can be used to initialized a ResolvedFrame.
+
+    Returns:
+        new_schedule: A new schedule where frames have been replaced with
+            their corresponding Drive, Control, and/or Measure channels.
+
+    Raises:
+        PulseError: if a frame is not configured.
+    """
+    if frames_config is None:
+        return schedule
+
+    resolved_frames = {}
+    sample_duration = None
+    for frame_settings in frames_config.values():
+        frame = ResolvedFrame(**frame_settings)
+        frame.set_frame_instructions(schedule)
+        resolved_frames[frame.index] = frame
+        sample_duration = frame_settings['sample_duration']
+
+    if sample_duration is None:
+        raise PulseError('Frame configuration does not have a sample duration.')
+
+    # Used to keep track of the frequency and phase of the channels
+    channel_trackers = {}
+    for ch in schedule.channels:
+        if isinstance(ch, chans.PulseChannel):
+            channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    # Add the channels that the frames broadcast on.
+    for frame in resolved_frames.values():
+        for ch in frame.channels:
+            if ch not in channel_trackers:
+                channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    sched = Schedule(name=schedule.name, metadata=schedule.metadata)
+
+    for time, inst in schedule.instructions:
+        if isinstance(inst, instructions.Play):
+            chan = inst.channel
+
+            if isinstance(inst.operands[0], Signal):
```

Comment:
Do you think `operands[0]` is more clear to understand than `pulse`?

Reply:
No. But I found `if isinstance(inst.pulse, Signal)` a bit funny to read but its still okay so I changed to `pulse` in a1b160d .